### DISCUSSION TITLE
Pushable Physics Objects

### DISCUSSION THREAD

**Discussion by MudkipWithShades** [Techokami/SonicWorldsNext#113] [Q&A]
I haven't figured out my problem, but it's best I ask here anyway. I created a pushable block and watched videos about RigidBody2D physics because they require a little bit of coding for the player. I went ahead and did that. However, during testing, the block would not push normally like certain blocks seen in classic Sonic games. The block would push in an odd way, and Sonic's pushing animation slightly flickers as he tries to push the block.
![pushable block nodes](https://github.com/Techokami/SonicWorldsNext/assets/126309909/c3b19a38-de0c-4b89-a810-aecafb4195f8)
![pushing block problem](https://github.com/Techokami/SonicWorldsNext/assets/126309909/26547b47-75e3-4165-8fbc-64cc65ebd443)

**Comment by DimensionWarped** [comment]
Falling blocks in Sonic games are too simplistic for simulated physics objects to be a sensible approach to them. Pushing is always a tricky problem because your object is going to stop the motion of the player. Might be best to create a separate ground (Normal) state for pushing that the player enters when they start pushing and control the animation and speed of the player until the block either falls or the player does something that would cause the push state to end (changes/releases direction, jumps, etc)

**Reply by MudkipWithShades** [reply]
I kinda get what you mean, but I still don't understand. Do I create a new layer or something for the object so that way the player can interact with it, without involving too much code?

**Reply by DimensionWarped** [reply]
I'm honestly not thinking it needs to be a new layer so much as the player just needs to change how it interacts with terrain when pushing the wall starts and until the player stops pushing for some reason.

Basically think about it like this... the player starts in the normal state, the player starts pushing what is effectively a wall (IE the pushable object). The pushable objects should continue to be treated as a wall, but the player's state should change to something where we can programmatically control the player's X-Movement so that the velocity in the direction that the player is holding is pretty much set to the x-velocity of the object being pushed rather than using the normal rules for setting X-Speed. Then if for some reason the player is no longer positioned to push the object or they jump or something you return the player to normal.

For an example of something *vaguely* similar, the Carnival Night Barrel gimmick takes away the player's X *and* Y control and just has the barrel control everything. I'm thinking something *like* that, but not quite as constricting. Might be bet to get @Renhoex's opinions on this though.

**Comment by JimmyThrush** [comment]
My implementation of this is still unfinished, but my solution for the pushing block problem was to check (with "func physics_collision(body, hitVector") if whether the player is using the "push" animation, then, if so, check the direction from which the player is colliding. Then, move both the player and the block forward a certain amount.

Screencap of the script:
![10971240330985](https://github.com/Techokami/SonicWorldsNext/assets/123826402/66f829dd-b78f-400b-abbc-9399cc0332bd)

Completely unfinished, as you can see, and this approach probably has some problems, but the effect created is satisfactory so far.
[test.webm](https://github.com/Techokami/SonicWorldsNext/assets/123826402/7c4a2dc5-4a67-4729-861c-a725c3d8faa4)
